[GTK][WPE] Fix ASan clang build

philn opened the pull request:
#### ed8de7bc88561ffca7290537179ea80c76431fcc
<pre>
[GTK][WPE] Fix ASan clang build
<a href="https://bugs.webkit.org/show_bug.cgi?id=276592">https://bugs.webkit.org/show_bug.cgi?id=276592</a>

Reviewed by Michael Catanzaro.

The invalid-offsetof warning in Vector.h should be ignored for GCC and clang on linux, not only for
GCC.

* Source/WTF/wtf/Vector.h:
(WTF::VectorBuffer::endOfBuffer):

Canonical link: <a href="https://commits.webkit.org/280978@main">https://commits.webkit.org/280978@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/2319421b928708be05f4c53b2973864b992445fb

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/58226 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/37554 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/10709 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/61851 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/8671 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/45190 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/8865 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/47189 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 wincairo-tests](https://ews-build.webkit.org/#/builders/60/builds/6199 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/60257 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/35207 "Passed tests") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/50345 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/28023 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/31970 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/7642 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/7675 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/51318 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/53925 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/7911 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/63556 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/57469 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/2140 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/62/builds/7971 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [⏳ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/GTK-WK2-Tests-EWS "Waiting to run tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/2148 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/50357 "Passed tests") | [⏳ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/API-Tests-GTK-EWS "Waiting to run tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/12873 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/1835 "Passed tests") | | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/79229 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/33383 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/13170 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/34469 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/35553 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/34214 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

mcatanzaro commented on Source/WTF/wtf/Vector.h:
```diff
@@ -565,10 +565,13 @@ class VectorBuffer : private VectorBufferBase<T, Malloc> {
         ASSERT_WITH_SECURITY_IMPLICATION(buffer());
 
-        IGNORE_GCC_WARNINGS_BEGIN("invalid-offsetof")
+#if OS(LINUX)
```
Why only for Linux and not also for e.g. FreeBSD? Surely it's OK to always ignore the warning on any OS?

philn replied:
Didn't think of FreeBSD :)
I thought if we ignore this on every OS and for every compiler, why do we have that code in the first place?

I guess it's useful on Apple platforms?

mcatanzaro replied:
The static assert is still compiled on every platform. Only the compiler warning is ignored. And that warning should be suppressed everywhere.

The result of the assertion is undefined behavior, but since it's only a compile time assert, I think it's OK to keep the assert. We'd just have to remove or disable it on platforms where it fails to build.

philn replied:
OK I can remove the OS ifdef then, thanks :)

mcatanzaro commented on Source/WTF/wtf/Compiler.h:
```diff
@@ -456,14 +456,15 @@
 #endif
 
-#define IGNORE_WARNINGS_BEGIN(warning) IGNORE_WARNINGS_BEGIN_IMPL(GCC, warning)
-#define IGNORE_WARNINGS_END IGNORE_WARNINGS_END_IMPL(GCC)
```
What's the goal of this? Now we no longer benefit from suppressing -Wpragmas.

philn replied:
`IGNORE_WARNINGS_BEGIN` and its counterpart would expand to either the GCC or clang impl, depending on the selected compiler.

Before this change it would expand to the GCC impl and be useless when clang is the selected compiler.

mcatanzaro replied:
No: Clang respects the GCC pragmas and they are emitted unconditionally. I don't think the change is doing anything.

(I did misread the code, though. I see -Wpragmas is still suppressed, so this isn't hurting anything either.)

philn replied:
> Clang respects the GCC pragmas and they are emitted unconditionally. 

Why do you think I am sending this PR?
With the current main, clang and ASan enabled the warnings DO show up.

mcatanzaro replied:
They are showing up because the current code uses `IGNORE_GCC_WARNINGS_BEGIN`/`END` which indicates a desire to not ignore the warning if building with Clang.

All you need to do is change that to `IGNORE_WARNINGS_BEGIN`/`END` and then it will work as you desire.

mcatanzaro replied:
So what I'm saying is your changes in Vector.h should be sufficient.

`pragma GCC` is respected by both GCC and Clang.

`pragma clang` is respected by Clang only.

Switching to `pragma clang` when building with Clang, while harmless, also doesn't actually solve anything, since if building with Clang they're both equivalent.

philn replied:
Understood, thanks!

philn committed:
[GTK][WPE] Fix ASan clang build
https://bugs.webkit.org/show_bug.cgi?id=276592

Reviewed by Michael Catanzaro.

The invalid-offsetof warning in Vector.h should be ignored for GCC and clang on linux, not only for
GCC.

* Source/WTF/wtf/Vector.h:
(WTF::VectorBuffer::endOfBuffer):

Canonical link: https://commits.webkit.org/280978@main
